ledger/participant-integration-api/src/main/scala/platform/apiserver/update/UpdateMapper.scala

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.apiserver.update
+
+import com.daml.ledger.api.domain.{ParticipantParty, User}
+import com.daml.ledger.participant.state.index.v2._
+import com.daml.lf.data.Ref
+import com.google.protobuf.field_mask.FieldMask
+import com.daml.platform.apiserver.update.UpdatePathsTrie.MatchResult
+
+object UpdateMapper {
+
+  private val userAnnotationsPath = List(
+    FieldNames.UpdateUserRequest.user,
+    FieldNames.User.metadata,
+    FieldNames.Metadata.annotations,
+  )
+  private val userPrimaryPartyPath =
+    List(FieldNames.UpdateUserRequest.user, FieldNames.User.primaryParty)
+  private val userIsDeactivatePath =
+    List(FieldNames.UpdateUserRequest.user, FieldNames.User.isDeactivated)
+
+  private val fullUpdateUserTree = UpdatePathsTrie
+    .fromPaths(
+      Seq(
+        UpdatePath(userAnnotationsPath, modifier = UpdatePathModifier.NoModifier),
+        UpdatePath(userPrimaryPartyPath, modifier = UpdatePathModifier.NoModifier),
+        UpdatePath(userIsDeactivatePath, modifier = UpdatePathModifier.NoModifier),
+      )
+    )
+    .getOrElse(sys.error("Failed to create full update user tree. This should never happen"))
+
+  private def noUpdate[A]: Result[Option[A]] = Right(None)
+
+  /** Validates its input and produces an update object.
+    * NOTE: The return update object might represent an empty (no-op) update.
+    *
+    * @param user represents the new values for the update
+    * @param updateMask indicates which fields should get updated
+    */
+  def toUserUpdate(user: User, updateMask: FieldMask): Result[UserUpdate] = {
+
+    def validatePathsMatchValidFields(paths: Seq[UpdatePath]): Result[Unit] =
+      paths.foldLeft[Result[Unit]](Right(())) { (ax, parsedPath) =>
+        for {
+          _ <- ax
+          _ <-
+            if (!fullUpdateUserTree.containsPrefix(parsedPath.fieldPath)) {
+              Left(UpdatePathError.UnknownFieldPath(parsedPath.toRawString))
+            } else Right(())
+        } yield ()
+      }
+    def resolveAnnotationsUpdate(updateTrie: UpdatePathsTrie): Result[Option[AnnotationsUpdate]] =
+      updateTrie
+        .findMatch(userAnnotationsPath)
+        .fold(noUpdate[AnnotationsUpdate])(updateMatch =>
+          makeAnnotationsUpdate(newValue = user.metadata.annotations, updateMatch = updateMatch)
+        )
+    def resolvePrimaryPartyUpdate(updateTrie: UpdatePathsTrie): Result[Option[Option[Ref.Party]]] =
```

Comment:
I have not caught it earlier, but it does look a bit odd. Is this because the first option signifies existence of a path and the second signifies the setting of the primary party to null in case of None?

Reply:
The general type is `Result[Option[updateValueType]]` where the 'Option' tells us whether there should be an update or not.

For the primaryParty update, because it's an optional attribute, the 'updateValueType' is `Option[Ref.Party]`.

For the annotations update, the 'updateValueType' is is either `AnnotationsUpdate.Merge` or `AnnotationsUpdate.Replace`